presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java

```diff
@@ -454,6 +491,44 @@ public HashAggregationBuilder getAggregationBuilder()
         return aggregationBuilder;
     }
 
+    private void processInputPage(Page page)
+    {
+        if (preGroupedHashStrategy.isPresent()) {
+            // Running in segmented aggregation mode.
+            int lastRowInPage = page.getPositionCount() - 1;
+            if (currentSegment == null
+                    || preGroupedHashStrategy.get().rowEqualsRow(0, currentSegment.extractChannels(preGroupedChannels), lastRowInPage, page.extractChannels(preGroupedChannels))) {
+                // If this is the first page or the current segment might have more data in the incoming pages, process the whole page.
+                unfinishedWork = aggregationBuilder.processPage(page);
+            }
+            else if (preGroupedHashStrategy.get().rowEqualsRow(0, currentSegment.extractChannels(preGroupedChannels), 0, page.extractChannels(preGroupedChannels))) {
+                // If the current page starts with a new segment, flush before processing it.
+                remainingPage = page;
+            }
+            else {
+                // If the current segment ends in the current page, flush it with all the segments (if exist) except the last segment of the current page.
+                int lastSegmentStart = findLastSegmentStart(preGroupedHashStrategy.get(), page.extractChannels(preGroupedChannels));
+                unfinishedWork = aggregationBuilder.processPage(page.getRegion(0, lastSegmentStart));
+                remainingPage = page.getRegion(0, lastRowInPage - lastSegmentStart + 1);
```

Comment:
shouldn't the remainingPage be `page.getRegion(lastSegmentStart, lastRowInPage - lastSegmentStart + 1);`?

Reply:
I think this indicates that our test coverage is not enough and didn't catch this issue, we need to enhance our test coverage

Reply:
I just tested it out again, looks like our correctness test only tests one page which is why this error is not caught
Let's add more tests to test multiple pages and include different scenarios we discussed above

On top of my mind (but not limited to):
Test1: page 1 [1, 1, 1......1, 1], page 2 [2, 2, 2...2, 2], page 3 [2, 2, 3...3, 3]
Test2: page 1 [1, 1, 1......1, 2], page 2 [2, 2, 2...2, 3], page 3 [3, 3, 3...3, 3]
Test3: page 1 [1, 1, 1......1, 1], page 2 [1, 1, 1...1, 1], page 3 [1, 1, 1...1, 1]

Reply:
Added all the tests I can think of so far. The correctness will be well captured.